### DISCUSSION TITLE
`// @inject-into script`

### DISCUSSION THREAD

**Discussion by xBZZZZ** [violentmonkey/violentmonkey#1442] [Ideas]
`// @inject-into script` injects userscript in page global scope like html `<script>` tag

`unsafeWindow` and `GM*` functions and objects are not available so `// @grant` is ignored

edit: `// @require` is also ignored

example:
```javascript
// ==UserScript==
// @name        aaa
// @namespace   bbb
// @description ccc
// @author      ddd
// @include     /^https\:\/\/example\.com\/([?#]|$)/
// @version     0.1
// @grant       none
// @run-at      document-start
// @inject-into script
// ==/UserScript==

// logs `true, m`
let variableInPage="m";
console.log(`${window===this}, ${window.eval("variableInPage")}`);
```

**Comment by tophf** [comment]
AFAICT you can already do it by using GM_addElement.

**Reply by tophf** [reply]
Also, assuming the idea is to inject the entire script, how's this different to `@inject-into page` + `@grant none` ?

**Reply by xBZZZZ** [reply]
> Also, assuming the idea is to inject the entire script, how's this different to `@inject-into page` + `@grant none` ?

`@inject-into page` + `@grant none` injected script not into global scope but inside `(define,module,exports)=>{...}` (that is not page global scope)
```javascript
window.VMin2jvmu3nh61221ua7=function(VMin2jvmu3nh61221ua7){try{with(this)((define,module,exports)=>{// ==UserScript==
// @name        aaa
// @namespace   bbb
// @description ccc
// @author      ddd
// @include     /^https\:\/\/example\.com\/([?#]|$)/
// @version     0.1
// @grant       none
// @run-at      document-start
// @inject-into page
// ==/UserScript==

// this logs `true, m`
let variableInPage="m";
console.log(`${window===this}, ${window.eval("variableInPage")}`);
})()}catch(e){VMin2jvmu3nh61221ua7(e)}};0
```

**Reply by tophf** [reply]
I see the difference. So it seems like the only reason is to access let/const in the global scope, in which case GM_addElement should be sufficient?

**Reply by tophf** [reply]
Actually, there is already a key for this feature: [@unwrap](https://sourceforge.net/p/greasemonkey/wiki/Metadata_Block/#unwrap) so we can as well implement it for I see no reason not to.

**Reply by tophf** [reply]
PR: https://github.com/violentmonkey/violentmonkey/pull/1443

**Comment by gera2ld** [comment]
May I confirm that this tag is only used for intentionally exposing all top level variables to the global scope? Is there any real-world use case?

**Reply by tophf** [reply]
I *guess* this may be also useful for people who augmented sites via extensions like [user-javascript-and-css](https://chrome.google.com/webstore/detail/user-javascript-and-css/nbhcbdghjpllgmfilhnhkllmkecfmpld) that run the code as is, not as a userscript. There are several other extensions like this. This tag will allow them to migrate to VM painlessly.
